### ISSUE TITLE
FileNotFoundError in ensemble.py - Potential Fix Suggested

### ISSUE THREAD

**Issue by Kiri0824** [TakuyaFujimura/unlabeledasd#1]
Thank you for the clear and well-organized codebase! I've been working with it and encountered an issue when running `jobs/exp/exp2.sh`. It resulted in the following error:
`FileNotFoundError: Error: 0 files found for ../results/exp/dcase2024/featex_0256_4096_5/ToyCircuit/infer/version_epoch_12_14_16/ensemble_test_score.csv`

I checked the ensemble/ directory and noticed the presence of ensemble.sh. However, running it produced the same error.
After investigating, I believe the issue might be in ensemble/ensemble.py, line 44. Specifically, updating this line to:
`infer_dir.mkdir(parents=True, exist_ok=True) `resolved the problem for me.

With this change, the version_epoch_12_14_16/ensemble_test_result.csv and related files were successfully generated. 😊

Please let me know if this fix is appropriate or if additional steps are needed.

**Comment by TakuyaFujimura** [comment]
Thank you very much for your interest!
I sincerely apologize for any inconvenience caused by the codes.
I will check the issue as soon as possible.
Just a quick note to thank you for bringing this issue to my attention.

**Comment by Kiri0824** [comment]
> Thank you very much for your interest! I sincerely apologize for any inconvenience caused by the codes. I will check the issue as soon as possible. Just a quick note to thank you for bringing this issue to my attention.

Thank you for your quick reply. Please let me know if you have any new findings.😊

**Comment by Kiri0824** [comment]
> Thank you very much for your interest! I sincerely apologize for any inconvenience caused by the codes. I will check the issue as soon as possible. Just a quick note to thank you for bringing this issue to my attention.

I also found a FileNotFoundError issue. After executing `jobs/exp/exp1.sh` and `exp2.sh`, in the dcase2023 and dcase2024 folders under my result folder, there are folders like: **subloss_0256_4096_1**. However, in the **README.md of pseudoattr**, it is stated that:
For DCASE2023, `pattr_macdom_subloss_0256_4096_${seed}` should be executed.
For DCASE2024, `subloss_0256_4096_${seed}` should be executed.
These experiments are included in `jobs/exp/exp1.sh`.
`cd jobs/exp and execute ./exp1.sh.`

I'm not sure if it is appropriate to **change the lines 40 and 42 in pseudoattr/bash.sh from**: `--model_path "../results/exp/{dcase}/pattr_macdom_subloss_0256_4096_{seed}"` to `--model_path "../results/exp/{dcase}/subloss_0256_4096_{seed}"` when I execute `pseudoattr/pseudo_attr.sh`.

**Comment by Kiri0824** [comment]
> > Thank you very much for your interest! I sincerely apologize for any inconvenience caused by the codes. I will check the issue as soon as possible. Just a quick note to thank you for bringing this issue to my attention.
> 
> I also found a FileNotFoundError issue. After executing `jobs/exp/exp1.sh` and `exp2.sh`, in the dcase2023 and dcase2024 folders under my result folder, there are folders like: **subloss_0256_4096_1**. However, in the **README.md of pseudoattr**, it is stated that: For DCASE2023, `pattr_macdom_subloss_0256_4096_${seed}` should be executed. For DCASE2024, `subloss_0256_4096_${seed}` should be executed. These experiments are included in `jobs/exp/exp1.sh`. `cd jobs/exp and execute ./exp1.sh.`
> 
> I'm not sure if it is appropriate to **change the lines 40 and 42 in pseudoattr/bash.sh from**: `--model_path "../results/exp/{dcase}/pattr_macdom_subloss_0256_4096_{seed}"` to `--model_path "../results/exp/{dcase}/subloss_0256_4096_{seed}"` when I execute `pseudoattr/pseudo_attr.sh`.

also change the **lines 83 in pseudoattr/bash.sh** from: `--ckpt_path "embed/${dcase}/pattr_macdom_subloss_0256_4096_${seed}/interval_epoch=13-*"` to `--ckpt_path "embed/${dcase}/subloss_0256_4096_${seed}/interval_epoch=13-*"`.

**Comment by Kiri0824** [comment]
> > > Thank you very much for your interest! I sincerely apologize for any inconvenience caused by the codes. I will check the issue as soon as possible. Just a quick note to thank you for bringing this issue to my attention.
> > 
> > 
> > I also found a FileNotFoundError issue. After executing `jobs/exp/exp1.sh` and `exp2.sh`, in the dcase2023 and dcase2024 folders under my result folder, there are folders like: **subloss_0256_4096_1**. However, in the **README.md of pseudoattr**, it is stated that: For DCASE2023, `pattr_macdom_subloss_0256_4096_${seed}` should be executed. For DCASE2024, `subloss_0256_4096_${seed}` should be executed. These experiments are included in `jobs/exp/exp1.sh`. `cd jobs/exp and execute ./exp1.sh.`
> > I'm not sure if it is appropriate to **change the lines 40 and 42 in pseudoattr/bash.sh from**: `--model_path "../results/exp/{dcase}/pattr_macdom_subloss_0256_4096_{seed}"` to `--model_path "../results/exp/{dcase}/subloss_0256_4096_{seed}"` when I execute `pseudoattr/pseudo_attr.sh`.
> 
> also change the **lines 83 in pseudoattr/bash.sh** from: `--ckpt_path "embed/${dcase}/pattr_macdom_subloss_0256_4096_${seed}/interval_epoch=13-*"` to `--ckpt_path "embed/${dcase}/subloss_0256_4096_${seed}/interval_epoch=13-*"`.

I'm terribly sorry. I still can't execute jobs/exp/exp3.sh correctly. There are too many files missing. 😥

**Comment by TakuyaFujimura** [comment]
I'm sorry for the inconvenience. I appreciate your reporting.
This is an answer to the first issue (FileNotFoundError: …/ensemble_test_score.csv).
If you have any questions, please feel free to contact me.
Please give me a moment to check the second issue 🙇‍♂️

---

I guess the following sequence of events causes this problem:

1. `python ensemble.py ${cfg_path} --base_dir ${method_dir} --machines "${machines**[@]**}"`  (ensemble/base.sh, line 19) failed
    1. `…/infer/version_epoch_12_14_16` existed but `ensemble_test_score.csv` did not (I don’t know why)
    2. `exist_ok` is `False` (ensemble/ensemble.py, line 44)
    → Error
2. `python ../evaluate23.py --score_path "${score_path}" --all_mode` (ensemble/base.sh, line 25) causes FileNotFoundError: `…/ensemble_test_score.csv`

I’ve set exist_ok to False to prevent unexpected overwrites, but it’s okay to change it. (I’ll reflect it to the repository)

I still don't know why `…/infer/version_epoch_12_14_16` existed without `ensemble_test_score.csv` (maybe some problems have occurred in the `ensemble/ensemble.py`), but your change is fine.

**Comment by TakuyaFujimura** [comment]
@Kiri0824 
Thank you again for reporting issues. This is an answer to 2nd issue.

**TL;DR**

- I found that the README.md included some incorrect procedures. Therefore, I fixed the explanation on `README.md` and `pseudoattr/README.md`, and I split `jobs/exp/exp3.sh` into `jobs/exp/exp3_???.sh`
- Please pull the repository
- Please execute `pattr_bic_pre_subloss_0256_4096` included in `jobs/exp/exp3_PL.sh` after executing `jobs/exp/exp3_NA23.sh`.
- Your changes on `--model_path in pseudoattr/bash.sh` and `--ckpt_path in pseudoattr/bash.sh` are not correct (This is due to my code. My apologies.)

**Detailed explanation**

First, let me clarify the names of the methods. This is a correspondence table between the names in the experiment and the names in the paper.

  |  Name@paper   |  2023                                               |   2024                                            | 
  | ---           | ---                                                 | ---                                               | 
  |  N/A          |  pattr_macdom_subloss_0256_4096 (exp3_NA23.sh)      |  subloss_0256_4096 (exp1.sh)                      | 
  |  Class        |  pattr_bic_pre_subloss_0256_4096 (exp3_PL.sh)       |  pattr_bic_pre_subloss_0256_4096 (exp3_PL.sh)     | 
  |  Triplet      |  pattr_bic_triplet_subloss_0256_4096 (exp3_PL.sh)   |  pattr_bic_triplet_subloss_0256_4096 (exp3_PL.sh) | 
  |  PANNs        |  pattr_bic_panns_subloss_0256_4096 (exp3_PL.sh)     |  pattr_bic_panns_subloss_0256_4096 (exp3_PL.sh)   | 
  |  OpenL3       |  pattr_bic_openl3_subloss_0256_4096 (exp3_PL.sh)    |  pattr_bic_openl3_subloss_0256_4096 (exp3_PL.sh)  | 
  |  GT           |  subloss_0256_4096 (exp1.sh)                        |  pattr_gt_subloss_0256_4096 (exp3_GT24.sh)        | 



In the DCASE 2023, since all machines had the attribute labels, I used pseudo-labels that included only `machine-domain` information (i.e., `pattr_macdom`) for N/A, and I handled `subloss_0256_4096` that used original labels, as GT.
In the DCASE 2024, since some machines didn’t have attribute labels, I handled `subloss_0256_4096` as N/A, and I used ground-truth labels for GT.

---

Next, I will introduce how to execute the systems with pseudo-labels.

First, we need to generate pseudo-labels by executing `pseudoattr/pseudo_attr.sh` **after the preparations written in `pseudoattr/README.md`**

For PANNs and OpenL3, the script generates the pseudo-labels by using `pseudoattr/Cnn14_mAP=0.431.pth` and `torchopenl3` included in `unlabeledasd/pseudoattr/venv_ext`, respectively.

For Triplet, we need to construct a triplet learning-based feature extractor from scratch by executing `./exp3_triplet.sh` beforehand.

Class utilizes the feature extractor of the pre-trained N/A model (i.e., `pattr_macdom_subloss_0256_4096` in DCASE2023 and `subloss_0256_4096` in DCASE2024).

The previous `pseudoattr/README.md` had explained that these `pattr_macdom_subloss_0256_4096 in DCASE2023` and `subloss_0256_4096 in DCASE2024` are executed through `jobs/exp/exp1.sh`, but this explanation was **wrong**.

Although `subloss_0256_4096 (in DCASE2024)` was included in `jobs/exp/exp1.sh`, but `pattr_macdom_subloss_0256_4096 (in DCASE2023)` was not.

To solve this problem, I split `jobs/exp/exp3.sh` into `jobs/exp/exp3_GT24.sh` , `jobs/exp/exp3_NA23.sh` , and `jobs/exp/exp3_PL.sh` , and I modified `pseudoattr/README.md` as follows:

> ## Class
> - In DCASE2023, Class requires that `pattr_macdom_subloss_0256_4096_${seed}` has already been executed.
>     - `pattr_macdom_subloss_0256_4096_${seed}` is included in `jobs/exp/exp3_NA23.sh`. Please execute it first.
> - In DCASE2024, Class requires that `subloss_0256_4096_${seed}` has already been executed.
>     -  `subloss_0256_4096_${seed}` is included in `jobs/exp/exp1.sh`. Please execute it first.
> 

Therefore, please execute `pattr_bic_pre_subloss_0256_4096` included in `jobs/exp/exp3_PL.sh` after executing `jobs/exp/exp3_NA23.sh`.


If you have any questions, please feel free to ask me🙇‍♂️
